google-ads-examples/src/main/java/com/google/ads/googleads/examples/planning/GenerateKeywordIdeas.java

```diff
@@ -0,0 +1,198 @@
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param languageId the language ID.
+   * @param locationIds the location IDs.
+   * @param keywords the list of keywords to use as a seed for ideas.
+   * @param pageUrl optional URL related to your business to use as a seed for ideas.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   * @throws IllegalArgumentException if {@code keywords} is empty and {@code pageUrl} is null.
+   * @throws Exception if the example failed due to other errors.
+   */
+  private void runExample(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      long languageId,
+      List<Long> locationIds,
+      List<String> keywords,
+      @Nullable String pageUrl) {
+    try (KeywordPlanIdeaServiceClient keywordPlanServiceClient =
+        googleAdsClient.getKeywordPlanIdeaServiceClient()) {
+      GenerateKeywordIdeasRequest.Builder requestBuilder =
+          GenerateKeywordIdeasRequest.newBuilder()
+              .setCustomerId(Long.toString(customerId))
+              // Set the language resource using the provided language ID.
+              .setLanguage(StringValue.of(ResourceNames.languageConstant(languageId)))
+              // Set the network. To restrict to only Google Search, change the parameter below to
+              // KeywordPlanNetwork.GOOGLE_SEARCH.
+              .setKeywordPlanNetwork(KeywordPlanNetwork.GOOGLE_SEARCH_AND_PARTNERS);
+
+      // Add the resource name of each location ID to the request.
+      for (Long locationId : locationIds) {
+        requestBuilder.addGeoTargetConstants(
+            StringValue.of(ResourceNames.geoTargetConstant(locationId)));
+      }
+
+      // Make sure that keywords and/or page URL were specified. The request must have exactly one
+      // of urlSeed, keywordSeed, or keywordAndUrlSeed set.
+      if (keywords.isEmpty() && pageUrl == null) {
+        throw new IllegalArgumentException(
+            "At least one of keywords or page URL is required, but neither was specified.");
+      }
+
+      if (keywords.isEmpty()) {
+        // Only page URL was specified, so use a UrlSeed.
+        requestBuilder.getUrlSeedBuilder().setUrl(StringValue.of(pageUrl));
+      } else if (pageUrl == null) {
+        // Only keywords were specified, so use a KeywordSeed.
+        requestBuilder
+            .getKeywordSeedBuilder()
+            .addAllKeywords(keywords.stream().map(StringValue::of).collect(Collectors.toList()));
```

Comment:
Duplicated code for the stream().map(), maybe make a local variable?

Reply:
Unfortunately, if I do that I'd want to move back to the 2nd level of nesting. With the current nesting layout, creating the local variable is unnecessary for the `if` case. I'm concerned that it will just confuse readers.

I like the current nesting you suggested, so I'd prefer to keep it with the tradeoff of a bit of duplication. Also, the duplication means the `else if` and `else` blocks are a bit more amenable to copy/pasting.

@AnashOommen , what do you think?

Reply:
I agree with Josh, I prefer the ease of copy-pasting over a bit of duplication here.